src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,302 @@
+import * as _ from 'lodash';
+import * as grpc from 'grpc';
+import { EventEmitter } from 'events';
+
+import {
+	createInsecureCredentials,
+	createOrgIdMetadata,
+	Direction,
+	EntryAdapter,
+	PusherClient,
+	PushRequest,
+	PushResponse,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { sbvrUtils } from '@resin/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = sbvrUtils;
+
+const MAX_RETRIES = 10;
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function sleep(ms: number) {
+	return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: PusherClient;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = new PusherClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		return new Promise((resolve, reject) => {
```

Comment:
As a general point I prefer to put the `new Promise` around the absolute minimum code to show precisely what is being wrapped and can resolve/reject, eg
```ts
new Promise((resolve, reject) => {
	call.on('error', (error: Error & { details: string }) => {
		reject(...);
	});
	call.on('end', () => {
		resolve(...);
	});
});
```

Reply:
True, I have updated and its easier to read.